modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencyMessageValidator.scala

```diff
@@ -66,6 +72,25 @@ object CurrencyMessageValidator {
               case None => Validated.validNec(message)
             }
 
+        def validateIfAddressAlreadyUsed(
+          metagraphId: Address,
+          existingFeesAddresses: Map[Address, List[Address]],
+          message: Signed[CurrencyMessage]
+        ) = {
+          val metagraphAddresses = existingFeesAddresses.getOrElse(metagraphId, List.empty)
+          val messageAddress = message.address
+          if (metagraphAddresses.contains(messageAddress)) {
+            message.validNec
+          } else {
+            val allAddresses = existingFeesAddresses.values.flatten.toList
+            if (allAddresses.contains(messageAddress)) {
+              AddressAlreadyInUse.invalidNec
+            } else {
+              message.validNec
+            }
+          }
+        }
+
```

Comment:
I'd like to follow up on my previous comment about storage and synchronization issues. You already have access to `GlobalSnapshotInfo` (via `LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo]`), so instead of having a separate storage for message addresses, you could just retrieve all the information from `GlobalSnapshotInfo` and have a single source of truth. Currency L0 and Global L0 both have access to such a storage.

```scala
               def validateIfAddressAlreadyUsed(
          metagraphId: Address,
          message: Signed[CurrencyMessage]
        ) =
          lastGlobalSnapshotStorage.getCombined.map {
            case Some((_, info)) =>
              info.lastCurrencySnapshots.collect {
                case (address, Right((_, info))) => address -> info.lastMessages.toList.flatMap(_.values.map(_.address).toList).toSet
              }
            case None => SortedMap.empty[Address, Set[Address]]
          }.map { metagraphAddresses =>
            val messageAddress = message.address

            if (metagraphAddresses.get(metagraphId).exists(_.contains(messageAddress))) {
              message.validNec
            } else {
              if (metagraphAddresses.exists { case (_, addresses) => addresses.contains(messageAddress) }) {
                AddressAlreadyInUse.invalidNec
              } else {
                message.validNec
              }
            }
          }
```

Reply:
👍🏼